token.go

```diff
@@ -29,6 +29,11 @@ type token interface {
 	Less(token) bool
 }
 
+type tokenInt64 interface {
```

Comment:
Have you considered renaming `murmur3Token` to `int64Token` instead? I don't see any difference in those two except the `String()` implementation.

Reply:
There is another (big!) difference between those tokens: `Hash` (sorry, if that is obvious). 

`tokenInt64` is inspired by my Java driver implementation. Cassandra originally supported tokens that were represented by bytes (not only 64-bit int). However, some time ago, Scylla was modified to only support tokens represented by 64-bit int. This `tokenInt64` interface was meant (speaking from perspective of Java driver) to represent this decision/abstraction and to future-proof code if new partitioners were introduced (those would also be represented as 64-bit ints). Therefore, renaming `murmur3Token` would not make sense as `murmur3Token` is just one variant of `tokenInt64` tokens.

Reply:
> There is another (big!) difference between those tokens: `Hash` (sorry, if that is obvious).

`Hash` is a method on the partitioner, not the token. So `murmur3Partitioner` and `scyllaCDCPartitioner` can implement `Hash` differently and still return `int64Token`.

> `tokenInt64` is inspired by my Java driver implementation. Cassandra originally supported tokens that were represented by bytes (not only 64-bit int). However, some time ago, Scylla was modified to only support tokens represented by 64-bit int.

My point is that there are two types to represent the same token type now. Why is it a single type? Both `murmur3Token` and `scyllaCDCToken` need to be comparable and both have to know how to compare to the other type. If you look at the definition of the `token` interface that's how a token type is defined - by the comparison function. We have Stringer there as we need to format the tokens sometimes for errors/debugging. In my opinion it doesn't matter if a token is produced by `murmur3Partitioner` or `scyllaCDCPartitioner` as you handle the tokens generated by both in the same way. Having the different Stringer implementation doesn't help anything. So you have the same token type (int64 with `<` as comparison) implemented twice.

Yes, I know there could theoretically be different type of tokens using `int64` or a different type using something else than `int64`. However, fundamentally you have always just a single token type in a token ring for given keyspace (because you need to be able to compare/sort all the tokens in the ring). Before this pull request it was true that we also had a single type per token ring in the driver's code. Now a new type is introduced.

Suppose that you add another partitioner that uses a different hash to compute int64 token value. You will still have to implement comparison so the type would look exactly the same as `murmur3Token` or `scyllaCDCToken` (as it has to be comparable with scyllaCDCToken, so will implement the same `Less` function). Therefore it will just add another code duplication without any benefit. You could have just returned `int64Token` from the new implementation of the partitioner.

If you wanted to add an incompatible token type that uses int64 internally, but a different comparison function, you still could do it (but it would be incompatible with scyllaCDCToken so you can't implement `AsInt64` for this incompatible type anyway). Because `AsInt64` method/`tokenInt64` interface just says that it should return an int64 token with `<` as comparison function.

Suppose that you add another partitioner that uses different type than int64, for example int128. So for a `some128Token` you will have to introduce a new matching `scyllaCDCToken128` type in order to be able to compare between those token types.

Therefore, in my opinion, `murmur3Token` and `scyllaCDCToken` are tightly coupled now even though there is an interface to cast them to int64 and they represent the same type (int64 tokens compared with `Less` implemented as `<`).

> This `tokenInt64` interface was meant (speaking from perspective of Java driver) to represent this decision/abstraction and to future-proof code if new partitioners were introduced (those would also be represented as 64-bit ints). Therefore, renaming `murmur3Token` would not make sense as `murmur3Token` is just one variant of `tokenInt64` tokens.

As I tried to explain above, this won't future-proof the code. Either the new token type will use int64 with `<` as `Less` comparison (and we can use the `murmur3Token` renamed to `int64Token`) or it won't (and you don't know what the token type would look like, it certainly won't implement the `tokenInt64` interface by definition because of the different comparison function). Adding new type just makes the code more complicated. We don't need to future-proof the code now (because you don't know how the requirements for the new token will look like anyway, if you knew, you could have add the new type directly right now). Adding code "that might be useful in the future" is a common mistake in software design. Most of the time you can't predict your future business requirements. Let's handle the supposed new token type only after we know how it will look like. I think we won't be adding it any time soon. Also, we are talking about types that are not exported outside of the package, so we can introduce the additional abstraction once needed (but not sooner, please).

Reply:
I think what @martin-sucha proposes makes sense. Merged `murmur3Token` and `scyllaCDCToken` into one, `int64Token` (I also changed the name from `tokenInt64` to `int64Token` because it follows the existing convention of `xyzToken` better).

If somebody wishes to differentiate those token types later and e.g. implement pretty-printing differnetly, I guess they can do it easily without duplicating token comparison logic, e.g. like this:

```go
// int64Token is exactly as in the current version of PR

type scyllaCDCToken int64Token

func (p scyllaCDCPartitioner) ParseString(str string) token {
    var d int64
    if _, err := fmt.Sscanf(str, "CDCToken(%d)", &d); err != nil {
        return scyllaCDCMinToken
    }
    return scyllaCDCToken(d)
}

func (s scyllaCDCToken) String() string {
    return fmt.Sprintf("CDCToken(%d)", int64(s))
}

// (scyllaCDCToken).Less() is "inherited" from int64Token
```